examples/transformer/model.py

```diff
@@ -0,0 +1,195 @@
+            self.train()
+        else:
+            self.eval()
+
+        # Source word embedding
+        src_word_embeds = self.word_embedder(encoder_input)
+        src_word_embeds = src_word_embeds * self.config_model.hidden_dim ** 0.5
+
+        # Position embedding (shared b/w source and target)
+        src_seq_len = torch.full(
+            (batch_size,), encoder_input.size(1), dtype=torch.int32
+        )
+        src_seq_len = src_seq_len.to(device=encoder_input.device)
+
+        src_pos_embeds = self.pos_embedder(sequence_length=src_seq_len)
+        src_input_embedding = src_word_embeds + src_pos_embeds
+
+        encoder_output = self.encoder(
+            inputs=src_input_embedding, sequence_length=encoder_input_length
+        )
+
+        if is_train_mode:
+            assert decoder_input is not None
+            assert labels is not None
+
+            tgt_word_embeds = self.word_embedder(decoder_input)
+            tgt_word_embeds = (
+                tgt_word_embeds * self.config_model.hidden_dim ** 0.5
+            )
+            tgt_seq_len = decoder_input.new_full(
+                (batch_size,), decoder_input.size(1),
+            )
+
+            tgt_pos_embeds = self.pos_embedder(sequence_length=tgt_seq_len)
+
+            tgt_input_embedding = tgt_word_embeds + tgt_pos_embeds
+
+            # For training
+            outputs = self.decoder(
+                memory=encoder_output,
+                memory_sequence_length=encoder_input_length,
+                inputs=tgt_input_embedding,
+                decoding_strategy="train_greedy",
+            )
+            label_lengths = (labels != 0).long().sum(dim=1)
+            is_target = (labels != 0).float()
+            mle_loss = self.smoothed_loss_func(
+                outputs.logits, labels, label_lengths
+            )
+            mle_loss = (mle_loss * is_target).sum() / is_target.sum()
+            return mle_loss
+
+        else:
+            start_tokens = encoder_input.new_full(
+                (batch_size,), self.vocab.bos_token_id,
+            )
+
+            def _embedding_fn(x, y):
+                word_embed = self.word_embedder(x)
+                scale = self.config_model.hidden_dim ** 0.5
+                pos_embed = self.pos_embedder(y)
+                return word_embed * scale + pos_embed
+
+            predictions = self.decoder(
+                memory=encoder_output,
+                memory_sequence_length=encoder_input_length,
+                beam_width=beam_width,
+                length_penalty=self.config_model.length_penalty,
+                start_tokens=start_tokens,
+                end_token=self.vocab.eos_token_id,
+                embedding=_embedding_fn,
+                max_decoding_length=self.config_data.max_decoding_length,
+                decoding_strategy="infer_greedy",
+            )
+            # Uses the best sample by beam search
+            return predictions
+
+
+class LabelSmoothingLoss(nn.Module):
```

Comment:
Is this necessary to be a class? In TF, it's a function

Reply:
TF functions are generally more powerful as they're only called once during model construction. This class carries a buffer (`one_hot`) so I think we can keep it as a class.

Also, PyTorch provides functional-style and class-style interfaces for a lot of components (e.g. `F.relu` vs. `torch.nn.ReLU`), so basically either style is okay.